# CAPSTONE PROJECT: SEGMENTING AND CLUSTERING NEIGHBORHOODS IN TORONTO

# 1. Create a dataframe with Postal Code

IMPORT LIBRARY

In [1]:
import numpy as np
import pandas as pd
import json as js
import csv

# convert an address into latitude and longitude values
!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

from bs4 import BeautifulSoup

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.

Libraries imported.


Retrieve data from the wikipedia Page of Toronto

In [2]:
#Scrap data from website
wikiUrl= 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
req=requests.get(wikiUrl)
soup = BeautifulSoup(req.content,'lxml')
table = soup.find_all('table')[0]
df = pd.read_html(str(table))
PostalCodedf=df[0]
#PostalCodedf.head(10)

#Drop the Borough not assigned & rename columns according to the request
PSCdf=PostalCodedf[PostalCodedf.Borough!='Not assigned'].reset_index(drop=True)
PSCdf=PSCdf.rename(columns={'Postal code':'PostalCode'})

print(len(PSCdf['PostalCode'].unique()))
print(len(PSCdf['PostalCode']))

print(PSCdf[PSCdf.Neighborhood =='Not assigned'])
print(PSCdf[PSCdf.Neighborhood ==''])


PSCdf.head()

103
103
Empty DataFrame
Columns: [PostalCode, Borough, Neighborhood]
Index: []
Empty DataFrame
Columns: [PostalCode, Borough, Neighborhood]
Index: []


PostalCode           Borough                                  Neighborhood
0        M3A        North York                                     Parkwoods
1        M4A        North York                              Victoria Village
2        M5A  Downtown Toronto                    Regent Park / Harbourfront
3        M6A        North York             Lawrence Manor / Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

Regarding the results, it seems that that no postal codes are listed twice, and no borough are not assigned.
Yet let's clean the dataframe in case the list could change.

In [3]:
PSCdf['Neighborhood'].groupby([PSCdf.PostalCode,PSCdf.Borough]).apply(', '.join).reset_index().sort_values(by='PostalCode')
PSCdf.loc[PSCdf.Neighborhood =='Not assigned', 'Neighborhood']=PSCdf.loc[PSCdf.Neighborhood =='Not assigned', 'Borough']
PSCdf

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
..         ...               ...   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                           Regent Park / Harbourfront  
3                    Lawrence Manor / Lawrence Heights  
4         Queen's Park / Ontario Provincial Government  
..                                                 ...  
98     The Kingsway / Montgomery Road / Old Mill North  
99                                Church and Wellesley  
100              Business reply mail Processing CentrE  
101  Old Mill South / King's Mill Park / Sunnylea /...  
102  Mimico NW / The Queensway West / South of Bloo...  

[103 rows x 3 columns]

In [4]:
print('The dataframe has {} boroughs and {} Postal Code.'.format(
        len(PSCdf['Borough'].unique()),
        PSCdf.shape[0]
    )
)

#Testdf=pd.DataFrame
#column_names = ['Borough', 'Neighborhood']
#Testdf=Testdf(columns=column_names)
#Testdf['Borough']=['Yes',"Yes","No","Yes"]
#Testdf['Neighborhood']=['Lauren',"Adrien","John","Marc"]

#Testdf.loc[Testdf.Borough=='No', "Borough"]=Testdf.loc[Testdf.Borough=='No', "Neighborhood"]
#Testdf

The dataframe has 10 boroughs and 103 Postal Code.


We cleaned the data with a result of a DataFrame of 103 rows ie 103 postal code

# 2 Retrieve Longitude and Latitude and insert into the DataFrame

In [5]:
## retrieve data from CSV file
#!wget -q -O 'Geospatial_data.csv''http://cocl.us/Geospatial_data'

csvurl='http://cocl.us/Geospatial_data'
#csvfile
location_df=pd.read_csv(csvurl)
#print('Data downloaded!')

#####Let's amened the column name of location_df
location_df=location_df.rename(columns={'Postal Code':'PostalCode'})
location_df.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

In [6]:
####Let's merge the two dataframes to have the final Postal Code dataframe with locations
PSC_df=pd.merge(PSCdf, location_df,on='PostalCode', how='inner')
PSC_df.head()
PSC_df

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
..         ...               ...   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                           Regent Park / Harbourfront  43.654260 -79.360636  
3                    Lawrence Manor / Lawrence Heights  43.718518 -79.464763  
4         Queen's Park / Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98     The Kingsway / Montgomery Road / Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100              Business reply mail Processing CentrE  43.662744 -79.321558  
101  Old Mill South / King's Mill Park / Sunnylea /...  43.636258 -79.498509  
102  Mimico NW / The Queensway West / South of Bloo...  43.628841 -79.520999  

[103 rows x 5 columns]

# 3. Explore and cluster the neighborhoods in Toronto

## Define Foursquare Credentials and Version - to check please enter yours

In [7]:
CLIENT_ID = 'MZYWVN1CNWKMO53KA2FK0UB4UJBMTBBHEYM2NPHRBMJWEH0L' # your Foursquare ID
CLIENT_SECRET = 'AP1UHS23RCV0SNY2LIAQKM1BZWXR5Q144JLZSM21WFVI5XS1' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

#### Redefine the dataframe to have only Toronto names in it

In [8]:
Toronto_data = PSC_df[PSC_df['Borough'].str.contains('Toronto')].reset_index(drop=True)
Toronto_data

PostalCode           Borough  \
0         M5A  Downtown Toronto   
1         M7A  Downtown Toronto   
2         M5B  Downtown Toronto   
3         M5C  Downtown Toronto   
4         M4E      East Toronto   
5         M5E  Downtown Toronto   
6         M5G  Downtown Toronto   
7         M6G  Downtown Toronto   
8         M5H  Downtown Toronto   
9         M6H      West Toronto   
10        M5J  Downtown Toronto   
11        M6J      West Toronto   
12        M4K      East Toronto   
13        M5K  Downtown Toronto   
14        M6K      West Toronto   
15        M4L      East Toronto   
16        M5L  Downtown Toronto   
17        M4M      East Toronto   
18        M4N   Central Toronto   
19        M5N   Central Toronto   
20        M4P   Central Toronto   
21        M5P   Central Toronto   
22        M6P      West Toronto   
23        M4R   Central Toronto   
24        M5R   Central Toronto   
25        M6R      West Toronto   
26        M4S   Central Toronto   
27        M5S  Downtown Toronto   
28        M6S      West Toronto   
29        M4T   Central Toronto   
30        M5T  Downtown Toronto   
31        M4V   Central Toronto   
32        M5V  Downtown Toronto   
33        M4W  Downtown Toronto   
34        M5W  Downtown Toronto   
35        M4X  Downtown Toronto   
36        M5X  Downtown Toronto   
37        M4Y  Downtown Toronto   
38        M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
0                          Regent Park / Harbourfront  43.654260 -79.360636  
1        Queen's Park / Ontario Provincial Government  43.662301 -79.389494  
2                            Garden District, Ryerson  43.657162 -79.378937  
3                                      St. James Town  43.651494 -79.375418  
4                                         The Beaches  43.676357 -79.293031  
5                                         Berczy Park  43.644771 -79.373306  
6                                  Central Bay Street  43.657952 -79.387383  
7                                            Christie  43.669542 -79.422564  
8                          Richmond / Adelaide / King  43.650571 -79.384568  
9                       Dufferin / Dovercourt Village  43.669005 -79.442259  
10  Harbourfront East / Union Station / Toronto Is...  43.640816 -79.381752  
11                          Little Portugal / Trinity  43.647927 -79.419750  
12                      The Danforth West / Riverdale  43.679557 -79.352188  
13          Toronto Dominion Centre / Design Exchange  43.647177 -79.381576  
14     Brockton / Parkdale Village / Exhibition Place  43.636847 -79.428191  
15                    India Bazaar / The Beaches West  43.668999 -79.315572  
16                    Commerce Court / Victoria Hotel  43.648198 -79.379817  
17                                    Studio District  43.659526 -79.340923  
18                                      Lawrence Park  43.728020 -79.388790  
19                                           Roselawn  43.711695 -79.416936  
20                                   Davisville North  43.712751 -79.390197  
21                           Forest Hill North & West  43.696948 -79.411307  
22                     High Park / The Junction South  43.661608 -79.464763  
23                                 North Toronto West  43.715383 -79.405678  
24              The Annex / North Midtown / Yorkville  43.672710 -79.405678  
25                            Parkdale / Roncesvalles  43.648960 -79.456325  
26                                         Davisville  43.704324 -79.388790  
27                    University of Toronto / Harbord  43.662696 -79.400049  
28                                Runnymede / Swansea  43.651571 -79.484450  
29                       Moore Park / Summerhill East  43.689574 -79.383160  
30        Kensington Market / Chinatown / Grange Park  43.653206 -79.400049  
31  Summerhill West / Rathnelly / South Hill / For...  43.686412 -79.400049  
32  CN Tower / King and Spad

In [9]:
print('The dataframe has now {} Postal Code.'.format(Toronto_data.shape[0]))

The dataframe has now 39 Postal Code.


### Explore Venues in Toronto

#### Define function to extract the category of the venue

In [10]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [11]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called *Toronto_venues*

In [12]:
LIMIT=100
Toronto_venues = getNearbyVenues(names=Toronto_data['Neighborhood'],
                                   latitudes=Toronto_data['Latitude'],
                                   longitudes=Toronto_data['Longitude']
                                  )

Regent Park / Harbourfront
Queen's Park / Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond / Adelaide / King
Dufferin / Dovercourt Village
Harbourfront East / Union Station / Toronto Islands
Little Portugal / Trinity
The Danforth West / Riverdale
Toronto Dominion Centre / Design Exchange
Brockton / Parkdale Village / Exhibition Place
India Bazaar / The Beaches West
Commerce Court / Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West
High Park / The Junction South
North Toronto West
The Annex / North Midtown / Yorkville
Parkdale / Roncesvalles
Davisville
University of Toronto / Harbord
Runnymede / Swansea
Moore Park / Summerhill East
Kensington Market / Chinatown / Grange Park
Summerhill West / Rathnelly / South Hill / Forest Hill SE / Deer Park
CN Tower / King and Spadina / Railway Lands / Harbourfront West / Bathurst  Quay / South Niagara / Island airport
Rosed

In [13]:
print(Toronto_venues.shape)
Toronto_venues.head()

(1635, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park / Harbourfront               43.65426              -79.360636   
1  Regent Park / Harbourfront               43.65426              -79.360636   
2  Regent Park / Harbourfront               43.65426              -79.360636   
3  Regent Park / Harbourfront               43.65426              -79.360636   
4  Regent Park / Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3     Body Blitz Spa East       43.654735       -79.359874   
4      Morning Glory Cafe       43.653947       -79.361149   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4       Breakfast Spot

In [14]:
Toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            55   
Brockton / Parkdale Village / Exhibition Place                         24   
Business reply mail Processing CentrE                                  19   
CN Tower / King and Spadina / Railway Lands / H...                     17   
Central Bay Street                                                     65   
Christie                                                               17   
Church and Wellesley                                                   74   
Commerce Court / Victoria Hotel                                       100   
Davisville                                                             34   
Davisville North                                                       11   
Dufferin / Dovercourt Village                                          15   
First Canadian Place / Underground city                               100   
Forest Hill North & West                                                4   
Garden District, Ryerson                                              100   
Harbourfront East / Union Station / Toronto Isl...                    100   
High Park / The Junction South                                         25   
India Bazaar / The Beaches West                                        20   
Kensington Market / Chinatown / Grange Park                            62   
Lawrence Park                                                           3   
Little Portugal / Trinity                                              43   
Moore Park / Summerhill East                                            3   
North Toronto West                                                     21   
Parkdale / Roncesvalles                                                14   
Queen's Park / Ontario Provincial Government                           31   
Regent Park / Harbourfront                                             47   
Richmond / Adelaide / King                                             97   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede / Swansea                                                    41   
St. James Town                                                         86   
St. James Town / Cabbagetown                                           44   
Stn A PO Boxes                                                         95   
Studio District                                                        41   
Summerhill West / Rathnelly / South Hill / Fore...                     16   
The Annex / North Midtown / Yorkville                                  22   
The Beaches                                                             5   
The Danforth West / Riverdale                                          43   
Toronto Dominion Centre / Design Exchange                             100   
University of Toronto / Harbord                                        35   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             55   
Brockton / Parkdale Village / Exhibition Place                          24   
Business reply mail Processing CentrE                                   19   
CN Tower / King and Spadina / Railway Lands / H...                      17   
Central Bay Street                                                      65   
Christie                                                                17   
Church and Wellesley                                                    74   
Commerce Court / Victoria Hotel                                        100   
Davisville                                                              34   
Davisville North                    

In [15]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 226 uniques categories.


###  Analyze Each Neighborhood

In [16]:
# one hot encoding
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Toronto_onehot['Neighborhood'] = Toronto_venues['Neighborhood'] 
Toronto_onehotInt=Toronto_onehot.drop(columns=['Neighborhood'])
fixed_columns=[Toronto_venues.columns[0]]+list(Toronto_onehotInt.columns[:])
#print(Toronto_venues['Neighborhood'] )

# move neighborhood column to the first column

Toronto_onehot = Toronto_onehot[fixed_columns]
Toronto_onehot.head()

Neighborhood  Airport  Airport Food Court  Airport Gate  \
0  Regent Park / Harbourfront        0                   0             0   
1  Regent Park / Harbourfront        0                   0             0   
2  Regent Park / Harbourfront        0                   0             0   
3  Regent Park / Harbourfront        0                   0             0   
4  Regent Park / Harbourfront        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  ...  Theme Restaurant  Toy / Game Store  Trail  \
0             0         0  ...                 0                 0      0   
1             0         0  ...                 0                 0      0   
2             0         0  ...                 0                 0      0   
3             0         0  ...                 0                 0      0   
4             0         0  ...                 0                 0      0   

   Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0              0                              0                 0   
1              0                              0                 0   
2              0                              0                 0   
3              0                              0                 0   
4              0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Women's Store  Yoga Studio  
0                      0         0              0            0  
1                      0         0              0            0  
2                      0         0              0            0  
3                      0         0              0            0  
4                      0         0              0            0  

[5 rows x 226 columns]

In [17]:
Toronto_onehot.shape

(1635, 226)

In [18]:
Toronto_grouped = Toronto_onehot.groupby('Neighborhood').mean().reset_index()
Toronto_grouped

Neighborhood   Airport  \
0                                         Berczy Park  0.000000   
1      Brockton / Parkdale Village / Exhibition Place  0.000000   
2               Business reply mail Processing CentrE  0.000000   
3   CN Tower / King and Spadina / Railway Lands / ...  0.058824   
4                                  Central Bay Street  0.000000   
5                                            Christie  0.000000   
6                                Church and Wellesley  0.000000   
7                     Commerce Court / Victoria Hotel  0.000000   
8                                          Davisville  0.000000   
9                                    Davisville North  0.000000   
10                      Dufferin / Dovercourt Village  0.000000   
11            First Canadian Place / Underground city  0.000000   
12                           Forest Hill North & West  0.000000   
13                           Garden District, Ryerson  0.000000   
14  Harbourfront East / Union Station / Toronto Is...  0.000000   
15                     High Park / The Junction South  0.000000   
16                    India Bazaar / The Beaches West  0.000000   
17        Kensington Market / Chinatown / Grange Park  0.000000   
18                                      Lawrence Park  0.000000   
19                          Little Portugal / Trinity  0.000000   
20                       Moore Park / Summerhill East  0.000000   
21                                 North Toronto West  0.000000   
22                            Parkdale / Roncesvalles  0.000000   
23       Queen's Park / Ontario Provincial Government  0.000000   
24                         Regent Park / Harbourfront  0.000000   
25                         Richmond / Adelaide / King  0.000000   
26                                           Rosedale  0.000000   
27                                           Roselawn  0.000000   
28                                Runnymede / Swansea  0.000000   
29                                     St. James Town  0.000000   
30                       St. James Town / Cabbagetown  0.000000   
31                                     Stn A PO Boxes  0.000000   
32                                    Studio District  0.000000   
33  Summerhill West / Rathnelly / South Hill / For...  0.000000   
34              The Annex / North Midtown / Yorkville  0.000000   
35                                        The Beaches  0.000000   
36                      The Danforth West / Riverdale  0.000000   
37          Toronto Dominion Centre / Design Exchange  0.000000   
38                    University of Toronto / Harbord  0.000000   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0             0.000000      0.000000        0.000000         0.000000   
1             0.000000      0.000000        0.000000         0.000000   
2             0.000000      0.000000        0.000000         0.000000   
3             0.058824      0.058824        0.117647         0.176471   
4             0.000000      0.000000        0.000000         0.000000   
5             0.000000      0.000000        0.000000         0.000000   
6             0.000000      0.000000        0.000000         0.000000   
7             0.000000      0.000000        0.000000         0.000000   
8             0.000000      0.000000        0.000000         0.000000   
9             0.000000      0.000000        0.000000         0.000000   
10            0.000000      0.000000        0.000000         0.000000   
11            0.000000      0.000000        0.000000         0.000000   
12            0.000000      0.000000        0.000000         0.000000   
13            0.000000      0.000000        0.000000         0.000000   
14            0.000000      0.000000        0.000000         0.000000   
15            0.000000      0.000000        0.000000         0.000000   
16            0.000000      0.000000        0.000000         0.000000   
17            0.000000      0.000000        0.

In [19]:
Toronto_grouped.shape

(39, 226)

#### Let's print each neighborhood along with the top 5 most common venues

In [20]:
num_top_venues = 5

for hood in Toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = Toronto_grouped[Toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.05
1  Seafood Restaurant  0.04
2      Farmers Market  0.04
3          Restaurant  0.04
4            Beer Bar  0.04


----Brockton / Parkdale Village / Exhibition Place----
            venue  freq
0            Café  0.12
1  Breakfast Spot  0.08
2     Coffee Shop  0.08
3       Nightclub  0.08
4          Bakery  0.04


----Business reply mail Processing CentrE----
                venue  freq
0  Light Rail Station  0.11
1         Yoga Studio  0.05
2       Garden Center  0.05
3         Pizza Place  0.05
4          Comic Shop  0.05


----CN Tower / King and Spadina / Railway Lands / Harbourfront West / Bathurst  Quay / South Niagara / Island airport----
              venue  freq
0   Airport Service  0.18
1    Airport Lounge  0.12
2  Airport Terminal  0.12
3           Airport  0.06
4               Bar  0.06


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.18
1  Italian Restaurant  0.06
2 

function to sort the venues in descending order

In [21]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood

In [22]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Toronto_grouped['Neighborhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1     Brockton / Parkdale Village / Exhibition Place                  Café   
2              Business reply mail Processing CentrE    Light Rail Station   
3  CN Tower / King and Spadina / Railway Lands / ...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0    Italian Restaurant            Restaurant              Beer Bar   
1             Nightclub        Breakfast Spot           Coffee Shop   
2           Yoga Studio         Auto Workshop            Skate Park   
3        Airport Lounge      Airport Terminal               Airport   
4    Italian Restaurant        Sandwich Place                  Café   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                  Café           Cheese Shop          Cocktail Bar   
1             Pet Store         Burrito Place            Restaurant   
2            Smoke Shop                   Spa         Burrito Place   
3       Harbor / Marina           Coffee Shop   Rental Car Location   
4          Burger Joint       Bubble Tea Shop           Salad Place   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                Bakery    Seafood Restaurant         Farmers Market  
1          Climbing Gym           Yoga Studio                    Bar  
2        Farmers Market  Fast Food Restaurant             Restaurant  
3      Sculpture Garden         Boat or Ferry                    Bar  
4   Japanese Restaurant                   Spa       Sushi Restaurant

## Cluster Neighborhoods in Toronto

In [23]:
# set number of clusters # after several tests I decided to go with 3 as K cluster
kclusters = 3

Toronto_grouped_clustering = Toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:39] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 2, 0, 2, 0,
       0, 0, 0, 0, 2, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [24]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Toronto_merged = Toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Toronto_merged = Toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

Toronto_merged.head(15)

PostalCode           Borough  \
0         M5A  Downtown Toronto   
1         M7A  Downtown Toronto   
2         M5B  Downtown Toronto   
3         M5C  Downtown Toronto   
4         M4E      East Toronto   
5         M5E  Downtown Toronto   
6         M5G  Downtown Toronto   
7         M6G  Downtown Toronto   
8         M5H  Downtown Toronto   
9         M6H      West Toronto   
10        M5J  Downtown Toronto   
11        M6J      West Toronto   
12        M4K      East Toronto   
13        M5K  Downtown Toronto   
14        M6K      West Toronto   

                                         Neighborhood   Latitude  Longitude  \
0                          Regent Park / Harbourfront  43.654260 -79.360636   
1        Queen's Park / Ontario Provincial Government  43.662301 -79.389494   
2                            Garden District, Ryerson  43.657162 -79.378937   
3                                      St. James Town  43.651494 -79.375418   
4                                         The Beaches  43.676357 -79.293031   
5                                         Berczy Park  43.644771 -79.373306   
6                                  Central Bay Street  43.657952 -79.387383   
7                                            Christie  43.669542 -79.422564   
8                          Richmond / Adelaide / King  43.650571 -79.384568   
9                       Dufferin / Dovercourt Village  43.669005 -79.442259   
10  Harbourfront East / Union Station / Toronto Is...  43.640816 -79.381752   
11                          Little Portugal / Trinity  43.647927 -79.419750   
12                      The Danforth West / Riverdale  43.679557 -79.352188   
13          Toronto Dominion Centre / Design Exchange  43.647177 -79.381576   
14     Brockton / Parkdale Village / Exhibition Place  43.636847 -79.428191   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0                0           Coffee Shop                  Park   
1                0           Coffee Shop                 Diner   
2                0           Coffee Shop        Clothing Store   
3                0                  Café           Coffee Shop   
4                0     Health Food Store      Asian Restaurant   
5                0           Coffee Shop    Italian Restaurant   
6                0           Coffee Shop    Italian Restaurant   
7                0         Grocery Store                  Café   
8                0           Coffee Shop                  Café   
9                0                Bakery              Pharmacy   
10               0           Coffee Shop              Aquarium   
11               0                   Bar            Restaurant   
12               0      Greek Restaurant           Coffee Shop   
13               0           Coffee Shop                 Hotel   
14               0                  Café             Nightclub   

   3rd Most Common Venue          4th Most Common Venue  \
0                    Pub                         Bakery   
1            Yoga Studio                     Hobby Shop   
2                   Café      Middle Eastern Restaurant   
3           Cocktail Bar                          Hotel   
4                  Trail                            Pub   
5             Restaurant                       Beer Bar   
6         Sandwich Place                           Café   
7                   Park                    Coffee Shop   
8             Restaurant                            Gym   
9            Pizza Place                           Café   
10                  Café                     Restaurant   
11      Asian Restaurant  Vegetarian / Vegan Restaurant   
12    Italian Restaurant                 Ice Cream Shop   
13                  Café                     Restaurant   
14        Breakfast Spot                    Coffee Shop   

        5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0          Mexican Restaurant        Breakfast Spot            Restaurant   
1               Burri

In [25]:
Toronto_grouped = pd.merge(Toronto_grouped, Toronto_merged[['Neighborhood','Cluster Labels']], on='Neighborhood', how='inner')
Toronto_grouped.head()
Toronto_grouped.drop(columns=['Neighborhood'])
Toronto_grouped.groupby('Cluster Labels').mean()

Airport  Airport Food Court  Airport Gate  Airport Lounge  \
Cluster Labels                                                              
0               0.00173             0.00173       0.00173         0.00346   
1               0.00000             0.00000       0.00000         0.00000   
2               0.00000             0.00000       0.00000         0.00000   

                Airport Service  Airport Terminal  American Restaurant  \
Cluster Labels                                                           
0                       0.00519           0.00346             0.011537   
1                       0.00000           0.00000             0.000000   
2                       0.00000           0.00000             0.000000   

                Antique Shop  Aquarium  Art Gallery  ...  Theme Restaurant  \
Cluster Labels                                       ...                     
0                   0.002112  0.001471     0.004922  ...          0.000397   
1                   0.000000  0.000000     0.000000  ...          0.000000   
2                   0.000000  0.000000     0.000000  ...          0.000000   

                Toy / Game Store     Trail  Train Station  \
Cluster Labels                                              
0                       0.000865  0.006566       0.000882   
1                       0.000000  0.000000       0.000000   
2                       0.000000  0.125000       0.000000   

                Vegetarian / Vegan Restaurant  Video Game Store  \
Cluster Labels                                                    
0                                    0.008258          0.001134   
1                                    0.000000          0.000000   
2                                    0.000000          0.000000   

                Vietnamese Restaurant  Wine Bar  Women's Store  Yoga Studio  
Cluster Labels                                                               
0                            0.004239  0.004141       0.000303     0.010231  
1                            0.000000  0.000000       0.000000     0.000000  
2                            0.000000  0.000000       0.000000     0.000000  

[3 rows x 225 columns]

#### Cluster 1

In [26]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               0           Coffee Shop   
1   Downtown Toronto               0           Coffee Shop   
2   Downtown Toronto               0           Coffee Shop   
3   Downtown Toronto               0                  Café   
4       East Toronto               0     Health Food Store   
5   Downtown Toronto               0           Coffee Shop   
6   Downtown Toronto               0           Coffee Shop   
7   Downtown Toronto               0         Grocery Store   
8   Downtown Toronto               0           Coffee Shop   
9       West Toronto               0                Bakery   
10  Downtown Toronto               0           Coffee Shop   
11      West Toronto               0                   Bar   
12      East Toronto               0      Greek Restaurant   
13  Downtown Toronto               0           Coffee Shop   
14      West Toronto               0                  Café   
15      East Toronto               0  Fast Food Restaurant   
16  Downtown Toronto               0           Coffee Shop   
17      East Toronto               0                  Café   
20   Central Toronto               0                  Park   
22      West Toronto               0                   Bar   
23   Central Toronto               0        Clothing Store   
24   Central Toronto               0        Sandwich Place   
25      West Toronto               0             Gift Shop   
26   Central Toronto               0        Sandwich Place   
27  Downtown Toronto               0                  Café   
28      West Toronto               0           Pizza Place   
30  Downtown Toronto               0                  Café   
31   Central Toronto               0                   Pub   
32  Downtown Toronto               0       Airport Service   
34  Downtown Toronto               0           Coffee Shop   
35  Downtown Toronto               0           Coffee Shop   
36  Downtown Toronto               0           Coffee Shop   
37  Downtown Toronto               0           Coffee Shop   
38      East Toronto               0    Light Rail Station   

   2nd Most Common Venue 3rd Most Common Venue          4th Most Common Venue  \
0                   Park                   Pub                         Bakery   
1                  Diner           Yoga Studio                     Hobby Shop   
2         Clothing Store                  Café      Middle Eastern Restaurant   
3            Coffee Shop          Cocktail Bar                          Hotel   
4       Asian Restaurant                 Trail                            Pub   
5     Italian Restaurant            Restaurant                       Beer Bar   
6     Italian Restaurant        Sandwich Place                           Café   
7                   Café                  Park                    Coffee Shop   
8                   Café            Restaurant                            Gym   
9               Pharmacy           Pizza Place                           Café   
10              Aquarium                  Café                     Restaurant   
11            Restaurant      Asian Restaurant  Vegetarian / Vegan Restaurant   
12           Coffee Shop    Italian Restaurant                 Ice Cream Shop   
13                 Hotel                  Café                     Restaurant   
14             Nightclub        Breakfast Spot                    Coffee Shop   
15                  Park                   Pub                 Sandwich Place   
16            Restaurant                  Café                          Hotel   
17           Coffee Shop             Gastropub                        Brewery   
20     Convenience Store        Sandwich Place              Food & Drink Shop   
22       Thai Restaurant    Mexican Restaurant                           Café   
23           Coffee Shop                  Park             Seafood Restaurant   
24                  Café           Coffee Shop              

#### Cluster 2

In [27]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
19  Central Toronto               1                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
19           Music Venue           Yoga Studio      Department Store   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
19  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
19   Dumpling Restaurant            Donut Shop       Doner Restaurant

#### Cluster 3

In [28]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
18   Central Toronto               2                  Park   
21   Central Toronto               2                  Park   
29   Central Toronto               2                  Park   
33  Downtown Toronto               2                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
18           Swim School              Bus Line          Dessert Shop   
21         Jewelry Store                 Trail      Sushi Restaurant   
29            Playground           Summer Camp         Deli / Bodega   
33            Playground                 Trail         Deli / Bodega   

   5th Most Common Venue        6th Most Common Venue  \
18  Ethiopian Restaurant            Electronics Store   
21      Department Store            Electronics Store   
29     Electronics Store  Eastern European Restaurant   
33     Electronics Store  Eastern European Restaurant   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
18  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   
21  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   
29          Dumpling Restaurant            Donut Shop      Doner Restaurant   
33          Dumpling Restaurant            Donut Shop      Doner Restaurant   

   10th Most Common Venue  
18       Doner Restaurant  
21       Doner Restaurant  
29                Dog Run  
33                Dog Run